aten/src/ATen/native/LegacyBridge.cpp

```diff
@@ -0,0 +1,469 @@
+#include <ATen/ATen.h>
+#include <ATen/NativeFunctions.h>
+#include <ATen/SparseTensorRef.h>
+#include <ATen/ExpandUtils.h>
+
+namespace at { namespace native {
+
+namespace {
+  // NB: Even though some of the functions we have ported are CUDA
+  // friendly, flipping the switch between native and non-native is
+  // an all or nothing affair, because the internal representation
+  // is different
+  static bool _has_native(const Tensor& self) {
+    return self.is_sparse() && !self.is_cuda();
+  }
+
+  static bool _type_has_native(const Type& dtype) {
+    return dtype.is_sparse() && !dtype.is_cuda();
+  }
+}
+
+// These native operations are not "really" native; they're actually just bridge
+// functions that decide whether or not to call native sparse functions, or
+// TH functions.  This file should be temporary; when all of TH gets ported, we
+// can just use the native mechanism straight.
+
+// TODO: Maybe the foo_ variants should call th_foo_
+
+Tensor norm(const Tensor & self, Scalar p) {
+  if (_has_native(self)) {
+    return native_norm(self, p);
+  } else {
+    return th_norm(self, p);
+  }
+}
+
+Tensor clone(const Tensor& self) {
+  if (_has_native(self)) {
+    return native_clone(self);
+  } else {
+    return th_clone(self);
+  }
+}
+
+Tensor& resize_as_(Tensor& self, const Tensor& the_template) {
+  if (_has_native(self)) {
+    return native_resize_as_(self, the_template);
+  } else {
+    return th_resize_as_(self, the_template);
+  }
+}
+
+Tensor& pow_out(Tensor& result, const Tensor& self, Scalar exponent) {
+  if (_has_native(self)) {
+    return native_pow_out(result, self, exponent);
+  } else {
+    return th_pow_out(result, self, exponent);
+  }
+}
+
+Tensor pow(const Tensor& self, Scalar exponent) {
+  if (_has_native(self)) {
+    return native_pow(self, exponent);
+  } else {
+    return th_pow(self, exponent);
+  }
+}
+
+Tensor& zero_(Tensor& self) {
+  if (_has_native(self)) {
+    return native_zero_(self);
+  } else {
+    return th_zero_(self);
+  }
+}
+
+// Note [CPU sparse is globally native]
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+// The current state of affairs is as follows:
+//
+//  - CPU sparse functionality is implemented natively
+//  - CUDA sparse functionality, and all other functionality, are implemented
+//    in TH.
+//
+// Thus, we need these trampoline functions, to help us decide whether or
+// not we can go to native implementations or not.  We expect the trampolines
+// to go away when things get ported to native for real.
+
+// Note [Multiple dispatch to sparse]
+// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+// In principle, there are some degrees of freedom in how we could gotten to our
+// native CPU sparse implementations.  In particular, the handling of how we are
+// simulating multiple dispatch is asymmetric with how th_add_out handles the
+// multiple dispatch.  We justify this in two ways: (1) knowing when to
```

Comment:
could you explain how it's asymmetric?

Reply:
```
// Note [Multiple dispatch to sparse]
// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// In an ideal world, we would use direct support for multiple dispatch to
// say that add(Dense, Dense) should dispatch to one function, while
// add(Dense, Sparse) should dispatch to another function.
//
// In a world where we only have single dispatch, we can single dispatch on
// the first function, and then do an is_sparse() test on the second argument
// to direct ourselves to the correct argument.
//
// We are in neither of those worlds.  Instead, we have a th_add function
// which has legacy implementations in the single dispatch world, BUT our
// actual add function needs to call s_native_add if the function *would have*
// utilized a kernel implemented in THS (CPU sparse tensors).
//
// th_add is "good old single dispatch" which internally handles the is_sparse()
// test and also handles broadcasting.  s_native_add works asymmetrically:
// it doesn't handle broadcasting at all, and it ASSUMES that the relevant
// argument is a CPU sparse tensor.  Why the asymmetry?  It turns out it is not
// so easy to figure out if a kernel is implemented in THS; it's not as simple
// as testing if the first argument is CPU and sparse, because, e.g.,
// in add(Dense, Sparse), the sparse kernel is in the second argument.  So,
// the trampoline function is going to know about the overloads *anyway*; it
// might as well also handle is_sparse() and broadcasting while it's at it.
//
// Why not change TH to follow this new scheme?  We could... but since it's
// all going away when we finish porting the TH functions to ATen, we haven't
```

New comment